sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedPartitions.scala

```diff
@@ -0,0 +1,281 @@
+    SortExec(_, _, left: QueryStageExec, _),
+    SortExec(_, _, right: QueryStageExec, _))
+      if supportOptimization(joinType, left, right) =>
+      val leftStats = getStatistics(left)
+      val rightStats = getStatistics(right)
+      val numPartitions = leftStats.bytesByPartitionId.length
+
+      val leftMedSize = medianSize(leftStats)
+      val rightMedSize = medianSize(rightStats)
+      logInfo(s"HandlingSkewedJoin left medSize: ($leftMedSize)" +
+        s" right medSize ($rightMedSize)")
+      logInfo(s"left bytes Max : ${leftStats.bytesByPartitionId.max}")
+      logInfo(s"right bytes Max : ${rightStats.bytesByPartitionId.max}")
+
+      val skewedPartitions = mutable.HashSet[Int]()
+      val subJoins = mutable.ArrayBuffer[SparkPlan](smj)
+      for (partitionId <- 0 until numPartitions) {
+        val isLeftSkew = isSkewed(leftStats, partitionId, leftMedSize)
+        val isRightSkew = isSkewed(rightStats, partitionId, rightMedSize)
+        val isSkewAndSupportsSplit =
+          (isLeftSkew && supportSplitOnLeftPartition(joinType)) ||
+            (isRightSkew && supportSplitOnRightPartition(joinType))
+
+        if (isSkewAndSupportsSplit) {
+          skewedPartitions += partitionId
+          val leftMapIdStartIndices = if (isLeftSkew && supportSplitOnLeftPartition(joinType)) {
+            estimateMapIdStartIndices(left, partitionId, leftMedSize)
+          } else {
+            Array(0)
+          }
+          val rightMapIdStartIndices = if (isRightSkew && supportSplitOnRightPartition(joinType)) {
+            estimateMapIdStartIndices(right, partitionId, rightMedSize)
+          } else {
+            Array(0)
+          }
+
+          for (i <- 0 until leftMapIdStartIndices.length;
+               j <- 0 until rightMapIdStartIndices.length) {
+            val leftEndMapId = if (i == leftMapIdStartIndices.length - 1) {
+              getShuffleStage(left).plan.shuffleDependency.rdd.partitions.length
+            } else {
+              leftMapIdStartIndices(i + 1)
+            }
+            val rightEndMapId = if (j == rightMapIdStartIndices.length - 1) {
+              getShuffleStage(right).
+                plan.shuffleDependency.rdd.partitions.length
+            } else {
+              rightMapIdStartIndices(j + 1)
+            }
+            // For the skewed partition, we set the id of shuffle query stage to -1.
+            // And skip this shuffle query stage optimization in 'ReduceNumShufflePartitions' rule.
+            val leftSkewedReader =
+              PostShufflePartitionReader(getShuffleStage(left).copy(id = -1),
+                partitionId, leftMapIdStartIndices(i), leftEndMapId)
+
+            val rightSkewedReader =
+              PostShufflePartitionReader(getShuffleStage(right).copy(id = -1),
+                partitionId, rightMapIdStartIndices(j), rightEndMapId)
+
+            subJoins +=
+              SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
```

Comment:
just for brainstorm: here we are joining 2 partitions, not 2 RDDs, so there will be no shuffle. Is it better to run hash join than SMJ? cc @maryannxue

Reply:
Furthermore, if only one side is skew, maybe better to plan a broadcast hash join instead of a union of many SMJs?